README.md
EngeMetadata package
================
Jason T. Serviss
2018-10-09

Release build:
<a href="https://travis-ci.org/EngeLab/EngeMetadata"><img src="https://travis-ci.org/EngeLab/EngeMetadata.svg?branch=master"></a>
Test coverage:
[![codecov](https://codecov.io/gh/EngeLab/EngeMetadata/branch/master/graph/badge.svg)](https://codecov.io/gh/EngeLab/EngeMetadata)

## Installation

``` r
if(!"devtools" %in% rownames(installed.packages())) {
  install.packages("devtools")
}
devtools::install_github("EngeLab/EngeMetadata")
```

## Template setup

All instructions herein are case sensitive and tense specific, i.e.
“Well” is not the same as “well” and “Well” is not the same as
“Wells”.

The template design is based on sheet precedence with Wells \> Columns
\> Plate. This means e.g. if a key is present in the Plate sheet and
also present in Columns sheet, the values in the Columns sheet will
overwrite the values in the Plate sheet where they differ.

### Workbook

  - Must include 3 sheets named “Plate”, “Columns”, and “Wells”. Note
    that the names are plural.

### Plate sheet

  - Should only contain 2 columns named “Key” and “Value”.
  - The unique\_key key is mandatory and should match the file name.
  - The wells\_in\_plate key is mandatory. Possible values are 384 or
    96. More possibilities in *Adding non-standard metadata* section.
  - Note that if the whole plate does not contain cells, this may be
    annotated with the “Missing” key (in either the Columns or Wells
    sheet). If “Missing” is given the value TRUE, it these samples will
    be removed from the metadata downstream. Otherwise the whole plate,
    according to the wells\_in\_plate key, is assumed to contain cells.

### Columns sheet

  - The Column key is mandatory. Note that this is singular.
  - The Column key values must include all columns present in the plate
    type given by the wells\_in\_plate key in the Plate sheet. Single
    digit columns should be prefixed with a 0, i.e. “01” not “1”.

### Wells sheet

  - The Well key is mandatory. Note that this is singular.
  - The Well key values must include all wells in the plate type given
    by the wells\_in\_plate key in the Plate sheet. Single digit columns
    should be prefixed with a 0, i.e. “A01” not “A1”.

### Dates

  - Dates are expected to have the word “date” somewhere in the key.
  - Dates including a month or day that is a single digit should be
    prefixed with a “0”, i.e. “01” for January and not “1” likewise “01”
    for the first of the month not “1”.
  - Dates should be formatted year, month, day without spaces.
    Acceptable syntax is: 20181201 or 181201.

## Importing metadata into R

There is currently only one exported function in the package which is
called “metadata”. The metadata function takes a plate/file name and a
googledrive path as arguments and downloads, processes, and returns the
metadata as a tibble. Precedence is resolved automatically and the
metadata is expanded so that there is one entry per sample.

## Adding non-standard metadata

There is some limited functionality for adding metadata for non-standard
datasets, e.g. datasets produced by other labs that don’t necessiarily
conform well to the typical Enge lab setup. This can be achieved by
substituting the value in the wells\_in\_plate key for the number of
samples and, in addition, removing all columns from the Columns sheet.
The current setup only allows for metadata associated with an entire
experiment or individual samples (which can be annotated in the wells
sheet).


NEWS
v0.1.2
  * Adding support for non-standard metadata. See README.
v0.1.1 
  * added functionality to remove all cells annotated with the Missing key = TRUE for plates that are not full

tests/testthat.R
library(testthat)
library(EngeMetadata)

test_check("EngeMetadata")


tests/testthat/test_metadata.R
context("metadata")

test_that("check that .processKeys outputs the expected result", {
  expect_identical(.processKeys(c("A.x", "A.y")), "A")
})

test_that("check that .layout outputs the expected result", {
  expect_true(is_tibble(.layout(384)))
  expect_true(is_tibble(.layout(96)))
})

test_that("check that .resolve works with nothing to resolve", {
  #setup input data
  wells <- pull(testData[[3]], wells_in_plate)
  layout <- .layout(wells)
  base <- map_dfr(1:nrow(layout), function(x) testData[[3]]) %>%
    bind_cols(layout, .)
  bind1 <- full_join(base, testData[[2]], by = "Column")
  keys1 <- .processKeys(colnames(bind1))
  
  #expected
  expected <- data.frame()
  
  #output
  output <- map_dfc(keys1,  ~.resolve(bind1, .x))
  
  #test
  expect_identical(expected, output)
})

test_that("check that .resolve works with something to resolve", {
  #setup input data
  wells <- pull(testData[[3]], wells_in_plate)
  layout <- .layout(wells)
  base <- map_dfr(1:nrow(layout), function(x) testData[[3]]) %>%
    bind_cols(layout, .)
  resolved1 <- full_join(base, testData[[2]], by = "Column")
  bind2 <- full_join(resolved1, testData[[1]], by = "Well")
  keys2 <- .processKeys(colnames(bind2))
  
  #expected
  expected.sc.mult <- 48L
  expected.sc.new <- 8L
  expected.sc.sng <- 120L
  expected.sc.NA <- 208L
  expected.oc.X <- 61L
  expected.oc.Y <- 96L
  expected.oc.Z <- 3L
  expected.oc.NA <- 224L
  
  #output
  output <- map_dfc(keys2,  ~.resolve(bind2, .x))
  output.sc <- table(output$Sort_condition, useNA = "ifany")
  output.oc <- table(output$other_condition, useNA = "ifany")
  
  #test
  expect_identical(unname(output.sc[1]), expected.sc.mult)
  expect_identical(unname(output.sc[2]), expected.sc.new)
  expect_identical(unname(output.sc[3]), expected.sc.sng)
  expect_identical(unname(output.sc[4]), expected.sc.NA)
  
  expect_identical(unname(output.oc[1]), expected.oc.X)
  expect_identical(unname(output.oc[2]), expected.oc.Y)
  expect_identical(unname(output.oc[3]), expected.oc.Z)
  expect_identical(unname(output.oc[4]), expected.oc.NA)
  
  expect_identical(.resolve(bind2, list()), bind2)
})

test_that("check that checkMeta errors when expected", {
  #check that unique_key exists
  tmp <- list(testData[[1]], testData[[2]], select(testData[[3]], -unique_key))
  expect_error(checkMeta(tmp), regexp = "unique_key key is missing from Plate sheet.")
  
  #check that file name matches unique key
  tmp <- testData
  names(tmp) <- rep("A", length(tmp))
  expect_error(checkMeta(tmp), regexp = "File name and unique_key do not match.")
  
  #check that the wells_in_plate variable is present in the Plate sheet
  tmp <- list(testData[[1]], testData[[2]], select(testData[[3]], -wells_in_plate))
  names(tmp) <- rep("test1", 3)
  expect_error(checkMeta(tmp), regexp = "wells_in_plate key missing from Plate sheet")
  
  #check that wells_in_plate is numeric
  tmp <- list(testData[[1]], testData[[2]], mutate(testData[[3]], wells_in_plate = "A"))
  names(tmp) <- rep("test1", 3)
  expect_error(checkMeta(tmp), "wells_in_plate key must equal 384 or 96 or be the number of samples")
  
  #check that the Column key in the Columns sheet is present.
  tmp <- list(testData[[1]], select(testData[[2]], -Column), testData[[3]])
  names(tmp) <- rep("test1", 3)
  expect_error(checkMeta(tmp), regexp = "The Column key in the Columns sheet is missing.")
  
  #check that the Column key in the Columns sheet is correct.
  tmp <- list(testData[[1]], mutate(testData[[2]], Column = c(NA, Column[-1])), testData[[3]])
  names(tmp) <- rep("test1", 3)
  expect_error(checkMeta(tmp), regexp = "The Column key in the Columns sheet is malformated.*")
  
  #check that the Well key in the Wells sheet is present.
  tmp <- list(select(testData[[1]], -Well), testData[[2]], testData[[3]])
  names(tmp) <- rep("test1", 3)
  expect_error(checkMeta(tmp), regexp = "The Well key in the Wells sheet is missing.")
  
  #check that the Wells key in the Wells sheet is correct.
  tmp <- list(mutate(testData[[1]], Well = c(NA, Well[-1])), testData[[2]], testData[[3]])
  names(tmp) <- rep("test1", 3)
  expect_error(checkMeta(tmp), regexp = "The Well key in the Wells sheet is malformated.*")
})

test_that("check that .datesFormat outputs the expected result", {
  expect_is(pull(testData[[3]], sample_recieve_date), "Date")
  
  #no date present
  tmp <- list(testData[[1]], testData[[2]], select(testData[[3]], -sample_recieve_date))
  expect_identical(tmp, .datesFormat(tmp))
  
  #malformatted dates
  tmp <- list(testData[[1]], testData[[2]], mutate(testData[[3]], sample_recieve_date = "2018102"))
  expect_warning(.datesFormat(tmp))
  
  tmp <- list(testData[[1]], testData[[2]], mutate(testData[[3]], sample_recieve_date = "2018-20-02"))
  expect_warning(.datesFormat(tmp))
})

test_that("check that resolvePlateMeta outputs the expected result", {
  output <- resolvePlateMeta(testData)
  expect_true(is_tibble(output))
  
  #check that all conflicts are resolved
  expect_false(all(str_detect(colnames(output), "\\.x$")))
  expect_false(all(str_detect(colnames(output), "\\.y$")))
  
  #check all columns present
  expected <- map(testData, colnames) %>% unlist() %>% unique()
  expect_true(all(expected %in% colnames(output)))
  
  #check dimensions
  expect_identical(nrow(output), 384L)
  expect_identical(ncol(output), 15L)
})

test_that("check that getPlateMeta outputs the expected result", {
  p <- system.file("testdata/test1.xlsx", package = "EngeMetadata")
  output <- getPlateMeta(
    plate = "test1", path = p, verbose = FALSE,
    safe = FALSE, local = TRUE
  )
  expect_is(output, "list")
  expect_identical(unique(names(output)), "test1")
  expect_true(all(unlist(map(output, is_tibble))))
  expect_identical(length(output), 3L)
})

test_that("check that .safe2 errors when expected", {
  p <- system.file("testdata/test1.xlsx", package = "EngeMetadata")
  expect_silent(.safe2(p))
  p <- system.file("testdata/test2.xlsx", package = "EngeMetadata")
  expect_error(.safe2(p))
})


tests/testthat/test-files/test1.R
#run from package root: source('tests/testthat/test-files/test1.R')
meta <- EngeMetadata:::getPlateMeta("test1", 'data/package_testing/')
save(meta, file = 'tests/testthat/test-files/test1.rda')


R/metadata.R
#' metadata
#'
#' Downloads and processes a metadata, formated according to a standard
#' template, for a specific plate and path on Google Drive.
#'
#' @name metadata
#' @rdname metadata
#' @param plate Character; Corresponds to the plate to process metadata for.
#' Must correspond to a file name in the path argument.
#' @param path Character; The location of the plate metadata on Google Drive.
#' @return A tibble with the metadata specified in the file.
#' @author Jason Serviss
#' @examples
#'
#' \dontrun{metadata("test1", 'Enge_lab/GFP_mouse/Annotation/package_testing')}
#'
NULL
#' @export

metadata <- function(
  plate,
  path = 'data/package_testing/',
  verbose = FALSE
){
  meta <- getPlateMeta(plate, path, verbose)
  checkMeta(meta)
  meta <- resolvePlateMeta(meta)
  return(meta)
}

#' getPlateMeta
#'
#' Downloads and reads the metadata into R.
#'
#' @name getPlateMeta
#' @keywords internal
#' @rdname getPlateMeta
#' @param plate Character; Corresponds to the plate to process metadata for.
#' Must correspond to a file name in the path argument.
#' @param path Character; The location of the plate metadata on Google Drive.
#' @param verbose Logical; Indicates if function should be verbose.
#' @param safe Logical; for internal use.
#' @param local Logical; for internal use. 
#' @return List; Length 3 list with the Wells, Columns, and Plate sheets,
#' respectivley.
#' @author Jason Serviss
NULL
#' @importFrom googledrive drive_download
#' @importFrom purrr map
#' @importFrom readxl read_excel
#' @importFrom tidyr spread
#' @importFrom stringr str_detect
#' @importFrom rlang .data

getPlateMeta <- function(
  plate, path = 'data/package_testing/', verbose, safe = TRUE, local = FALSE
){
  if(safe) .safe1(plate, path)

  #download plate data
  if(local) {
    p <- path
  } else {
    p <- file.path(tempdir(), paste0(plate, ".xlsx"))
    drive_download(plate, path = p, overwrite = TRUE, verbose = verbose)
  }
  
  if(safe) .safe2(p)
  
  #read and extract data
  meta <- map(c("Wells", "Columns", "Plate"), function(s) {
    d <- read_excel(path = p, sheet = s)
    if(s == "Plate") d <- spread(d, .data$Key, .data$Value, convert = TRUE)
    d
  })
  
  names(meta) <- rep(plate, 3)
  
  #format dates
  .datesFormat(meta)
}

#' .safe1
#'
#' Checks for file on googledrive before attempting to download.
#'
#' @name .safe1
#' @keywords internal
#' @rdname dot-safe1
#' @param plate Character; Corresponds to the plate to process metadata for.
#' Must correspond to a file name in the path argument.
#' @param path Character; The location of the plate metadata on Google Drive.
#' @author Jason Serviss
NULL
#' @importFrom googledrive drive_ls
#' @importFrom dplyr pull
#' @importFrom rlang .data

.safe1 <- function(plate, path) {
  files <- pull(drive_ls(path = path), .data$name)
  if(!plate %in% files) stop(paste0(plate, " not found on Google Drive."))
}

#' .safe2
#'
#' Checks that all sheets are present before loading the file.
#'
#' @name .safe2
#' @keywords internal
#' @rdname dot-safe2
#' @param path Character; The location of the plate metadata locally.
#' @author Jason Serviss
NULL
#' @importFrom readxl excel_sheets

.safe2 <- function(path) {
  if(!all(c("Plate", "Columns", "Wells") %in% excel_sheets(path))) {
    s <- c("Plate", "Columns", "Wells")
    missing <- s[which(!s %in% excel_sheets(path))]
    stop(paste0(missing, " is not present in the sheets of plate ", path))
  }
}

#' .datesFormat
#'
#' Date formatting helper to \code{\link{getPlateMeta}}.
#'
#' @name .datesFormat
#' @keywords internal
#' @rdname dot-datesFormat
#' @param meta Output from \code{\link{getPlateMeta}} function.
#' @return List of tibbles with ymd formatted dates.
#' @author Jason Serviss
NULL
#' @importFrom lubridate ymd
#' @importFrom purrr map map2
#' @importFrom stringr str_detect
#' @importFrom dplyr mutate sym
#' @importFrom rlang ":=" "!!"

.datesFormat <- function(meta) {
  dateCols <- map(meta, function(d) {
    colnames(d)[str_detect(colnames(d), "date")]
  })
  
  map2(meta, dateCols, function(x, y) {
    if(length(y) == 0) {x} else {
      mutate(x, !! y := ymd(!! dplyr::sym(y)))
    }
  })
}

#' checkMeta
#'
#' Checks that assumptions concerning the metadata in Google Drive are fufilled.
#'
#' @name checkMeta
#' @keywords internal
#' @rdname checkMeta
#' @param meta List; Length 3 list with the Wells, Columns, and Plate sheets,
#' respectivley.
#' @return Nothing.
#' @author Jason Serviss
NULL
#' @importFrom rlang .data

checkMeta <- function(meta) {
  #check that unique_key exists
  if(!"unique_key" %in% colnames(meta[[3]])) {
    stop("unique_key key is missing from Plate sheet.")
  }
  #check that file name matches unique key
  if(unique(names(meta)) != pull(meta[[3]], .data$unique_key)) {
    stop("File name and unique_key do not match.")
  }
  #check that the wells_in_plate variable is present in the Plate sheet
  if(!"wells_in_plate" %in% colnames(meta[[3]])) {
    stop("wells_in_plate key missing from Plate sheet")
  }
  #check that wells_in_plate is numeric
  if(!is.numeric(pull(meta[[3]], .data$wells_in_plate))) {
    stop("wells_in_plate key must equal 384 or 96 or be the number of samples")
  }
  
  #Only run column and well checks when the layout is 384 or 96
  wells <- pull(meta[[3]], .data$wells_in_plate)
  if(wells %in% c(384, 96)) {
    #check that the Column key in the Columns sheet is present.
    if(!"Column" %in% colnames(meta[[2]])) {
      stop("The Column key in the Columns sheet is missing.")
    }
    #check that the Column key in the Columns sheet is correct.
    wells <- pull(meta[[3]], .data$wells_in_plate)
    if(!identical(unique(.layout(wells)$Column), meta[[2]]$Column)) {
      mess <- paste0(
        "The Column key in the Columns sheet is malformated. ", 
        "The plate format is: ", wells,
        " so Column should contain: ", 
        paste(unique(.layout(wells)$Column), collapse = ", ")
      )
      stop(mess)
    }
    #check that the Well key in the Wells sheet is present.
    if(!"Well" %in% colnames(meta[[1]])) {
      stop("The Well key in the Wells sheet is missing.")
    }
    #check that the Wells key in the Wells sheet is correct.
    if(!identical(.layout(wells)$Well, meta[[1]]$Well)) {
      mess <- paste0(
        "The Well key in the Wells sheet is malformated. ", 
        "The plate format is ", wells,
        " so Well should contain: ", paste(.layout(wells)$Well, collapse = ", ")
      )
      stop(mess)
    }
  }
}

#' resolvePlateMeta
#'
#' Resolves the precedence of the of the keys in the metadata file with the
#' hierarchy Wells > Columns > Plates.
#'
#' @name resolvePlateMeta
#' @keywords internal
#' @rdname resolvePlateMeta
#' @param meta List; Length 3 list with the Wells, Columns, and Plate sheets,
#' respectivley. Typically from the \code{\link{getPlateMeta}} function.
#' @return A tibble with the resolved metadata.
#' @author Jason Serviss
NULL
#' @importFrom dplyr pull bind_cols full_join select matches "%>%"
#' @importFrom purrr map_dfr map_dfc
#' @importFrom tibble is_tibble
#' @importFrom rlang .data

resolvePlateMeta <- function(meta) {
  
  #layout plate
  wells <- pull(meta[[3]], .data$wells_in_plate)
  if(!wells %in% c(384, 96)) {
    ns <- .handleNonstandardLayout(meta, wells)
    return(ns)
  }
  layout <- .layout(wells)

  #Add plate data to layout
  base <- map_dfr(1:nrow(layout), function(x) meta[[3]]) %>%
    bind_cols(layout, .)
  
  #resolve Plate and Column prescedence
  bind1 <- full_join(base, meta[[2]], by = "Column")
  keys1 <- .processKeys(colnames(bind1))
  add1 <- map_dfc(keys1,  ~.resolve(bind1, .x))
  full1 <- bind_cols(bind1, add1)
  resolved1 <- select(full1, -(dplyr::matches("\\.[x-y]")))

  #resolve Well and combined Plate and Column prescedence
  bind2 <- full_join(resolved1, meta[[1]], by = "Well")
  keys2 <- .processKeys(colnames(bind2))
  add2 <- map_dfc(keys2,  ~.resolve(bind2, .x))
  full2 <- bind_cols(bind2, add2)
  out <- select(full2, -(dplyr::matches("\\.[x-y]")))
  
  return(out)
}

.handleNonstandardLayout <- function(meta, wells) {
  if(length(meta[[2]]) != 0) {
    stop("Column annotation detected in non-standard layout.")
  }
  base <- map_dfr(1:wells, function(x) meta[[3]])
  if(length(meta[[1]]) == 0) {
    return(base)
  }
  bind_cols(meta[[1]], base)
}

#' .processKeys
#'
#' Helper for \code{\link{resolvePlateMeta}}. Extracts duplicated key names with
#' a ".x" or ".y" suffix.
#'
#' @name .processKeys
#' @keywords internal
#' @param data A tibble with potentially dupicated keys in the colnames.
#' @return A character vector of the duplicated keys without their suffix.
#' @author Jason Serviss
NULL
#' @importFrom stringr str_replace

.processKeys <- function(keys) {
  kp <- str_replace(keys, "(.*)\\.[x-y]", "\\1")
  kp[duplicated(kp)]
}

#' .resolve
#'
#' Helper for \code{\link{resolvePlateMeta}}. Resolved key precedence.
#' Expects the greater suffix (i.e. .y > .x) to have greater precedence.
#' Expects only .x and .y suffixes.
#'
#' @name .resolve
#' @keywords internal
#' @param data Tibble; including the keys to be resolved.
#' @param key Character; the key to be resolved. key.x and key.y are expected
#' in colnames(data).
#' @return A character vector of the duplicated keys without their suffix.
#' @author Jason Serviss
NULL
#' @importFrom dplyr select mutate sym "%>%"
#' @importFrom rlang ":=" "!!"
#' @importFrom purrr map2
#' @importFrom tidyr unnest

.resolve <- function(data, key) {
  if(length(key) == 0) return(data)

  key.x <- paste0(key, ".x"); key.y <- paste0(key, ".y")
  key.x.sym <- sym(key.x); key.y.sym <- sym(key.y)

  if(!all(c(key.x, key.y) %in% colnames(data))) {
    stop("Keys missing from the data.")
  }

  data %>%
    select(key.x, key.y) %>%
    mutate(!! key := map2(!! key.y.sym, !! key.x.sym, function(y, x) {
      replace(y, is.na(y), x)
    })) %>%
    select(-key.x, -key.y) %>%
    unnest()
}

#' .layout
#'
#' Helper for \code{\link{resolvePlateMeta}}. Sets up the plate layout.
#'
#' @name .layout
#' @keywords internal
#' @param format Numeric; Indicates which plate to setup. Can be 384 or 96.
#' @return A tibble with the plate layout.
#' @author Jason Serviss
NULL
#' @importFrom tibble tibble is_tibble

.layout <- function(format) {
  Row <- NULL; Column <- NULL; Well <- NULL
  
  if(format == 384) {
    r <- LETTERS[1:16]
    c <- c(paste0("0", 1:9), 10:24)
  } else {
    r <- LETTERS[1:8]
    c <- c(paste0("0", 1:9), 10:12)
  }

  tibble(
    Row = rep(r, each = length(c)),
    Column = rep(c, length(r)),
    Well = paste0(Row, Column)
  )
}

